contributors/design-proposals/device-plugin.md

```diff
@@ -0,0 +1,557 @@
+  * deciding which device to assign to the pod's containers 
+  * Calling the `Allocate` function with the list of devices
+
+The scheduler is still in charge of filtering the nodes which cannot
+satisfy the resource requests.
+
+# Device Plugin
+
+## Introduction
+
+The device plugin is structured in 3 parts:
+1. Registration: The device plugin advertises it's presence to Kubelet
+2. ListAndWatch: The device plugin advertises a list of Devices to Kubelet
+   and sends it again if the state of a Device changes
+3. Allocate: When creating containers, Kubelet calls the device plugin's
+   `Allocate` function so that it can run device specific instructions (gpu
+    cleanup, QRNG initialization, ...) and instruct Kubelet how to make the
+    device available in the container.
+
+## Registration
+
+When starting the device plugin is expected to make a (client) gRPC call
+to the `Register` function that Kubelet exposes.
+
+The communication between Kubelet is expected to happen only through Unix
+sockets and follow this simple pattern:
+1. The device plugins sends a `RegisterRequest` to Kubelet (through a
+   gRPC request)
+2. Kubelet answers to the `RegisterRequest` with a `RegisterResponse`
+   containing any error Kubelet might have encountered
+3. The device plugin start it's gRPC server if it did not recieve an
+   error
+
+Note that in the register request, the device plugin lists the labels
+it wants the node to advertise.
+If the labels are conflicting with pre-existing labels, Kubelet does
+not answer with a successfull `RegisterRequest`.
+
+In the current alpha version, labels will be used as properties for
+the device.
+
+## Unix Socket
+
+Device Plugins are expected to communicate with Kubelet through gRPC
+on an Unix socket.
+When starting the gRPC server, they are expected to create a unix socket
+at the following host path: `/var/run/kubernetes`.
+
+For non bare metal device plugin this means they will have to mount the folder
+as a volume in their pod spec ([see Installation](##installation)).
+
+Device plugins can expect to find the socket to register themselves on
+the host at the following path:
+`/var/run/Kubernetes/kubelet.sock`.
+
+## Protocol Overview
+
+When first registering themselves against Kubelet, the device plugin
+will send:
+  * The name of their unix socket
+  * [The API version against which they were built](#versioning).
+  * Their `ResourceName` they want to advertise
+
+Kubelet answers with the minimum version it supports and whether or
+not there was an error. The errors may include (but not limited to):
+  * API version not supported
+  * A device plugin already registered this `ResourceName`
+
+Kubelet will then interact with the plugin through the following functions:
+  * ListAndWatch: The device plugin advertises a list of Devices to Kubelet
+    and sends it again if the state of a Device changes
+  * `Allocate`: Called when creating a container with a list of devices
+
+![Process](device-plugin.png)
```

Comment:
In the graph, maybe replace "Discovery Phase" with "Discovery/Monitoring Phase"?